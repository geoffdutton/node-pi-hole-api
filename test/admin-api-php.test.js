
const request = require('supertest')
const ftlResponses = require('./ftl-responses')
const app = require('../app')
const FTL = require('../api/pihole-ftl')

const base = '/admin/api.php'

describe(base, () => {
  beforeEach(() => {
    FTL.prototype.send = jest.fn((cmd, parser) => {
      let res

      if (ftlResponses.byCommand[cmd]) {
        const _res = ftlResponses.byCommand[cmd]
        res = parser ? parser(_res) : FTL.parseRawResponse(_res)
      } else {
        res = FTL.parseRawResponse(ftlResponses.stats)
      }

      return Promise.resolve(res)
    })
  })

  test('needs to include at least one approved query key', async () => {
    const response = await request(app).get(base + '?somethingNotAllowed')
    expect(response.body).toMatchSnapshot()
    expect(response.body.errorCode).toBe('InvalidMethod')
    expect(response.statusCode).toBe(400)
  })

  test('throws not implemented', async () => {
    const response = await request(app).get(base + '?__testNotImped')
    expect(response.body).toMatchSnapshot()
    expect(response.body.errorCode).toBe('MethodNotImplemented')
    expect(response.statusCode).toBe(500)
  })

  test('no query & ?summary', async () => {
    let response = await request(app).get(base)
    expect(response.statusCode).toBe(200)
    const body = response.body

    response = await request(app).get(base + '?summary')
    expect(response.statusCode).toBe(200)
    expect(body).toEqual(response.body)
    expect(response.body).toMatchSnapshot()
  })

  test('?summaryRaw', async () => {
    const response = await request(app).get(base + '?summaryRaw')
    expect(response.body).toMatchSnapshot()
    expect(response.statusCode).toBe(200)
  })

  test('?getClientNames', async () => {
    const response = await request(app).get(base + '?getClientNames')
    expect(response.body).toMatchSnapshot()
    expect(response.body.clients.length).toBe(5)
    expect(response.statusCode).toBe(200)
  })

  test('?overTimeData & ?overTimeData10mins', async () => {
    let response = await request(app).get(base + '?overTimeData')
    expect(response.statusCode).toBe(200)
    const body = response.body

    response = await request(app).get(base + '?overTimeData10mins')
    expect(response.statusCode).toBe(200)
    expect(body).toEqual(response.body)
    expect(response.body).toMatchSnapshot()
  })

  test('?type & ?version', async () => {
    let response = await request(app).get(base + '?type')
    expect(response.statusCode).toBe(200)
    const body = response.body

    response = await request(app).get(base + '?version')
    expect(response.statusCode).toBe(200)
    expect(body).toEqual(response.body)
    expect(response.body).toMatchSnapshot()
  })

  test('?topItems', async () => {
    const response = await request(app).get(base + '?topItems')
    expect(FTL.prototype.send).toHaveBeenCalledWith('top-ads')
    expect(FTL.prototype.send).toHaveBeenCalledWith('top-domains')
    expect(response.body.top_queries['play.hulu.com']).toBe(568)
    expect(response.body.top_ads['cws-hulu.conviva.com']).toBe(5676)
    expect(response.statusCode).toBe(200)
  })

  test('?topItems audit', async () => {
    const response = await request(app).get(base + '?topItems=audit')
    expect(FTL.prototype.send).toHaveBeenCalledWith('top-ads for audit')
    expect(FTL.prototype.send).toHaveBeenCalledWith('top-domains for audit')
    expect(response.body.top_queries['play.hulu.com']).toBe(568)
    expect(response.body.top_ads['cws-hulu.conviva.com']).toBe(5676)
    expect(response.statusCode).toBe(200)
  })

  test('?topItems 99', async () => {
    const response = await request(app).get(base + '?topItems=99')
    expect(FTL.prototype.send).toHaveBeenCalledWith('top-ads (99)')
    expect(FTL.prototype.send).toHaveBeenCalledWith('top-domains (99)')
    expect(response.body.top_queries['play.hulu.com']).toBe(568)
    expect(response.body.top_ads['cws-hulu.conviva.com']).toBe(5676)
    expect(response.statusCode).toBe(200)
  })

  test('?topClients & wih int', async () => {
    const cmds = ['topClients', 'getQuerySources']
    for (const cmd of cmds) {
      let response = await request(app).get(base + '?' + cmd)
      expect(response.body).toMatchSnapshot()
      expect(FTL.prototype.send).toHaveBeenCalledWith('top-clients')
      expect(response.body.top_sources['104.54.196.251|104-54-196-251.lightspeed.austtx.sbcglobal.net']).toBe(28830)
      const body = response.body

      response = await request(app).get(base + `?${cmd}=33`)
      expect(response.body).toMatchSnapshot()
      expect(body).toEqual(response.body)
      expect(FTL.prototype.send).toHaveBeenCalledWith('top-clients (33)')
    }
  })

  test('?topClientsBlocked & with int', async () => {
    let response = await request(app).get(base + '?topClientsBlocked')
    expect(response.body).toMatchSnapshot()
    expect(FTL.prototype.send).toHaveBeenCalledWith('top-clients blocked')
    expect(response.body.top_sources_blocked['104.54.196.251|104-54-196-251.lightspeed.austtx.sbcglobal.net']).toBe(16479)

    const body = response.body
    response = await request(app).get(base + '?topClientsBlocked=77')
    expect(FTL.prototype.send).toHaveBeenCalledWith('top-clients blocked (77)')
    expect(body).toEqual(response.body)
  })

  test('?getForwardDestinations & with unsorted', async () => {
    let response = await request(app).get(base + '?getForwardDestinations')
    expect(response.body).toMatchSnapshot()
    expect(FTL.prototype.send).toHaveBeenCalledWith('forward-dest')
    expect(response.body.forward_destinations['blocklist|blocklist']).toBe(56.12)

    const body = response.body
    response = await request(app).get(base + '?getForwardDestinations=unsorted')
    expect(FTL.prototype.send).toHaveBeenCalledWith('forward-dest unsorted')
    expect(body).toEqual(response.body)
  })

  test('?getForwardDestinationNames', async () => {
    let response = await request(app).get(base + '?getForwardDestinationNames')
    expect(response.body).toMatchSnapshot()
    expect(FTL.prototype.send).toHaveBeenCalledWith('forward-names')
    expect(response.body.forward_destinations['blocklist|blocklist']).toBe(56.12)
  })

  test('?getQueryTypes', async () => {
    const response = await request(app).get(base + '?getQueryTypes')
    expect(FTL.prototype.send).toHaveBeenCalledWith('querytypes')
    expect(response.body.querytypes['A (IPv4)']).toBe(48.45)
    expect(response.statusCode).toBe(200)
  })

  test('?getCacheInfo', async () => {
    const response = await request(app).get(base + '?getCacheInfo')
    expect(FTL.prototype.send).toHaveBeenCalledWith('cacheinfo')
    expect(response.body.cacheinfo['cache-live-freed']).toBe(0)
    expect(response.statusCode).toBe(200)
  })

  test('?overTimeDataQueryTypes', async () => {
    const response = await request(app).get(base + '?overTimeDataQueryTypes')
    expect(FTL.prototype.send).toHaveBeenCalledWith('QueryTypesoverTime')
    expect(response.body.over_time['1546056300']).toEqual([0, 0])
    expect(response.statusCode).toBe(200)
  })

  test('?overTimeDataClients', async () => {
    const response = await request(app).get(base + '?overTimeDataClients')
    expect(FTL.prototype.send).toHaveBeenCalledWith('ClientsoverTime')
    expect(response.body.over_time['1546056300']).toEqual([215, 3, 0, 0, 0, 0])
    expect(response.statusCode).toBe(200)
  })

  test('?recentBlocked', async () => {
    const response = await request(app).get(base + '?recentBlocked')
    expect(FTL.prototype.send).toHaveBeenCalledWith('recentBlocked', expect.any(Function))
    expect(response.headers['content-type']).toBe('text/plain; charset=utf-8')
    expect(response.statusCode).toBe(200)
    expect(response.text).toBe('vortex.hulu.com')
  })

  test('?getAllQueries', async () => {
    const response = await request(app).get(base + '?getAllQueries')
    expect(FTL.prototype.send).toHaveBeenCalledWith('getallqueries')
    expect(response.statusCode).toBe(200)
    expect(response.body.data[0]).toEqual([
      1546139804,
      'AAAA',
      'browsetech.slack.com',
      '104-54-196-251.lightspeed.austtx.sbcglobal.net',
      2,
      0,
      1,
      19
    ])
  })
})
