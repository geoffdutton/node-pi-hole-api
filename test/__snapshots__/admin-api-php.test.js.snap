// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/admin/api.php ?getClientNames 1`] = `
Object {
  "clients": Array [
    Object {
      "ip": "104.54.196.251",
      "name": "104-54-196-251.lightspeed.austtx.sbcglobal.net",
    },
    Object {
      "ip": "127.0.0.1",
      "name": "localhost",
    },
    Object {
      "name": "129.204.58.88",
    },
    Object {
      "name": "71.6.202.205",
    },
    Object {
      "name": "178.128.10.121",
    },
  ],
}
`;

exports[`/admin/api.php ?getForwardDestinationNames 1`] = `
Object {
  "forward_destinations": Object {
    "blocklist|blocklist": 56.12,
    "cache|cache": 9.88,
    "google-public-dns-a.google.com|8.8.8.8": 4.85,
    "google-public-dns-b.google.com|8.8.4.4": 5.58,
    "one.one.one.one|1.0.0.1": 9.97,
    "one.one.one.one|1.1.1.1": 13.54,
    "one.one.one.one|2606:4700:4700::1001": 0.06,
  },
}
`;

exports[`/admin/api.php ?getForwardDestinations & with unsorted 1`] = `
Object {
  "forward_destinations": Object {
    "blocklist|blocklist": 56.12,
    "cache|cache": 9.88,
    "google-public-dns-a.google.com|8.8.8.8": 4.85,
    "google-public-dns-b.google.com|8.8.4.4": 5.58,
    "one.one.one.one|1.0.0.1": 9.97,
    "one.one.one.one|1.1.1.1": 13.54,
    "one.one.one.one|2606:4700:4700::1001": 0.06,
  },
}
`;

exports[`/admin/api.php ?overTimeData & ?overTimeData10mins 1`] = `
Object {
  "ads_over_time": Object {
    "1546056300": 127,
    "1546056900": 100,
    "1546057500": 116,
    "1546058100": 104,
    "1546058700": 108,
    "1546059300": 116,
    "1546059900": 130,
    "1546060500": 141,
    "1546061100": 80,
    "1546061700": 106,
    "1546062300": 132,
    "1546062900": 128,
    "1546063500": 86,
    "1546064100": 84,
    "1546064700": 106,
    "1546065300": 92,
    "1546065900": 112,
    "1546066500": 122,
    "1546067100": 92,
    "1546067700": 112,
    "1546068300": 96,
    "1546068900": 126,
    "1546069500": 94,
    "1546070100": 104,
    "1546070700": 106,
    "1546071300": 104,
    "1546071900": 162,
    "1546072500": 102,
    "1546073100": 132,
    "1546073700": 96,
    "1546074300": 98,
    "1546074900": 122,
    "1546075500": 148,
    "1546076100": 128,
    "1546076700": 90,
    "1546077300": 114,
    "1546077900": 118,
    "1546078500": 96,
    "1546079100": 132,
    "1546079700": 120,
    "1546080300": 126,
    "1546080900": 100,
    "1546081500": 116,
    "1546082100": 82,
    "1546082700": 8,
    "1546083300": 0,
    "1546083900": 14,
    "1546084500": 38,
    "1546085100": 37,
    "1546085700": 25,
    "1546086300": 13,
    "1546086900": 90,
    "1546087500": 170,
    "1546088100": 111,
    "1546088700": 269,
    "1546089300": 230,
    "1546089900": 158,
    "1546090500": 167,
    "1546091100": 143,
    "1546091700": 144,
    "1546092300": 144,
    "1546092900": 138,
    "1546093500": 81,
    "1546094100": 142,
    "1546094700": 118,
    "1546095300": 126,
    "1546095900": 136,
    "1546096500": 118,
    "1546097100": 138,
    "1546097700": 116,
    "1546098300": 116,
    "1546098900": 132,
    "1546099500": 130,
    "1546100100": 134,
    "1546100700": 118,
    "1546101300": 120,
    "1546101900": 148,
    "1546102500": 108,
    "1546103100": 114,
    "1546103700": 130,
    "1546104300": 136,
    "1546104900": 118,
    "1546105500": 138,
    "1546106100": 118,
    "1546106700": 130,
    "1546107300": 163,
    "1546107900": 141,
    "1546108500": 148,
    "1546109100": 143,
    "1546109700": 169,
    "1546110300": 146,
    "1546110900": 150,
    "1546111500": 182,
    "1546112100": 139,
    "1546112700": 163,
    "1546113300": 144,
    "1546113900": 179,
    "1546114500": 138,
    "1546115100": 145,
    "1546115700": 168,
    "1546116300": 150,
    "1546116900": 151,
    "1546117500": 144,
    "1546118100": 150,
    "1546118700": 158,
    "1546119300": 140,
    "1546119900": 168,
    "1546120500": 160,
    "1546121100": 174,
    "1546121700": 152,
    "1546122300": 167,
    "1546122900": 150,
    "1546123500": 136,
    "1546124100": 152,
    "1546124700": 138,
    "1546125300": 146,
    "1546125900": 150,
    "1546126500": 187,
    "1546127100": 152,
    "1546127700": 145,
    "1546128300": 158,
    "1546128900": 138,
    "1546129500": 146,
    "1546130100": 48,
  },
  "domains_over_time": Object {
    "1546056300": 218,
    "1546056900": 139,
    "1546057500": 155,
    "1546058100": 146,
    "1546058700": 144,
    "1546059300": 160,
    "1546059900": 159,
    "1546060500": 190,
    "1546061100": 105,
    "1546061700": 148,
    "1546062300": 181,
    "1546062900": 169,
    "1546063500": 136,
    "1546064100": 113,
    "1546064700": 160,
    "1546065300": 127,
    "1546065900": 149,
    "1546066500": 169,
    "1546067100": 127,
    "1546067700": 159,
    "1546068300": 135,
    "1546068900": 183,
    "1546069500": 128,
    "1546070100": 152,
    "1546070700": 155,
    "1546071300": 155,
    "1546071900": 232,
    "1546072500": 144,
    "1546073100": 190,
    "1546073700": 123,
    "1546074300": 141,
    "1546074900": 164,
    "1546075500": 175,
    "1546076100": 180,
    "1546076700": 124,
    "1546077300": 164,
    "1546077900": 179,
    "1546078500": 129,
    "1546079100": 186,
    "1546079700": 157,
    "1546080300": 173,
    "1546080900": 135,
    "1546081500": 164,
    "1546082100": 163,
    "1546082700": 50,
    "1546083300": 5,
    "1546083900": 70,
    "1546084500": 243,
    "1546085100": 367,
    "1546085700": 124,
    "1546086300": 172,
    "1546086900": 277,
    "1546087500": 323,
    "1546088100": 387,
    "1546088700": 516,
    "1546089300": 435,
    "1546089900": 329,
    "1546090500": 378,
    "1546091100": 333,
    "1546091700": 321,
    "1546092300": 346,
    "1546092900": 283,
    "1546093500": 264,
    "1546094100": 268,
    "1546094700": 152,
    "1546095300": 158,
    "1546095900": 193,
    "1546096500": 162,
    "1546097100": 195,
    "1546097700": 146,
    "1546098300": 150,
    "1546098900": 189,
    "1546099500": 168,
    "1546100100": 182,
    "1546100700": 267,
    "1546101300": 150,
    "1546101900": 201,
    "1546102500": 126,
    "1546103100": 182,
    "1546103700": 162,
    "1546104300": 189,
    "1546104900": 160,
    "1546105500": 192,
    "1546106100": 146,
    "1546106700": 364,
    "1546107300": 399,
    "1546107900": 258,
    "1546108500": 350,
    "1546109100": 272,
    "1546109700": 334,
    "1546110300": 377,
    "1546110900": 274,
    "1546111500": 400,
    "1546112100": 242,
    "1546112700": 370,
    "1546113300": 313,
    "1546113900": 392,
    "1546114500": 289,
    "1546115100": 261,
    "1546115700": 412,
    "1546116300": 270,
    "1546116900": 302,
    "1546117500": 274,
    "1546118100": 300,
    "1546118700": 342,
    "1546119300": 256,
    "1546119900": 259,
    "1546120500": 292,
    "1546121100": 327,
    "1546121700": 301,
    "1546122300": 298,
    "1546122900": 235,
    "1546123500": 235,
    "1546124100": 267,
    "1546124700": 260,
    "1546125300": 295,
    "1546125900": 321,
    "1546126500": 338,
    "1546127100": 284,
    "1546127700": 258,
    "1546128300": 283,
    "1546128900": 253,
    "1546129500": 265,
    "1546130100": 83,
  },
}
`;

exports[`/admin/api.php ?summaryRaw 1`] = `
Object {
  "ads_blocked_today": 15116,
  "ads_percentage_today": 56.049538,
  "clients_ever_seen": 5,
  "dns_queries_all_types": 26969,
  "dns_queries_today": 26969,
  "domains_being_blocked": 136704,
  "privacy_level": 0,
  "queries_cached": 2706,
  "queries_forwarded": 9146,
  "reply_CNAME": 3778,
  "reply_IP": 11000,
  "reply_NODATA": 1552,
  "reply_NXDOMAIN": 85,
  "status": "enabled",
  "unique_clients": 5,
  "unique_domains": 654,
}
`;

exports[`/admin/api.php ?topClients & wih int 1`] = `
Object {
  "top_sources": Object {
    "104.54.196.251|104-54-196-251.lightspeed.austtx.sbcglobal.net": 28830,
    "127.0.0.1|localhost": 271,
    "129.204.58.88": 1,
    "178.128.10.121": 1,
    "71.6.202.205": 1,
    "74.82.47.22|scan-09e.shadowserver.org": 1,
  },
}
`;

exports[`/admin/api.php ?topClients & wih int 2`] = `
Object {
  "top_sources": Object {
    "104.54.196.251|104-54-196-251.lightspeed.austtx.sbcglobal.net": 28830,
    "127.0.0.1|localhost": 271,
    "129.204.58.88": 1,
    "178.128.10.121": 1,
    "71.6.202.205": 1,
    "74.82.47.22|scan-09e.shadowserver.org": 1,
  },
}
`;

exports[`/admin/api.php ?topClients & wih int 3`] = `
Object {
  "top_sources": Object {
    "104.54.196.251|104-54-196-251.lightspeed.austtx.sbcglobal.net": 28830,
    "127.0.0.1|localhost": 271,
    "129.204.58.88": 1,
    "178.128.10.121": 1,
    "71.6.202.205": 1,
    "74.82.47.22|scan-09e.shadowserver.org": 1,
  },
}
`;

exports[`/admin/api.php ?topClients & wih int 4`] = `
Object {
  "top_sources": Object {
    "104.54.196.251|104-54-196-251.lightspeed.austtx.sbcglobal.net": 28830,
    "127.0.0.1|localhost": 271,
    "129.204.58.88": 1,
    "178.128.10.121": 1,
    "71.6.202.205": 1,
    "74.82.47.22|scan-09e.shadowserver.org": 1,
  },
}
`;

exports[`/admin/api.php ?topClientsBlocked & with int 1`] = `
Object {
  "top_sources_blocked": Object {
    "104.54.196.251|104-54-196-251.lightspeed.austtx.sbcglobal.net": 16479,
    "127.0.0.1|localhost": 3,
  },
}
`;

exports[`/admin/api.php ?type & ?version 1`] = `
Object {
  "type": "FTL",
  "version": 3,
}
`;

exports[`/admin/api.php needs to include at least one approved query key 1`] = `
Object {
  "error": "Invalid Method Passed",
  "errorCode": "InvalidMethod",
  "query": Object {
    "somethingNotAllowed": "",
  },
}
`;

exports[`/admin/api.php no query & ?summary 1`] = `
Object {
  "ads_blocked_today": 15116,
  "ads_percentage_today": 56.049538,
  "clients_ever_seen": 5,
  "dns_queries_all_types": 26969,
  "dns_queries_today": 26969,
  "domains_being_blocked": 136704,
  "privacy_level": 0,
  "queries_cached": 2706,
  "queries_forwarded": 9146,
  "reply_CNAME": 3778,
  "reply_IP": 11000,
  "reply_NODATA": 1552,
  "reply_NXDOMAIN": 85,
  "status": "enabled",
  "unique_clients": 5,
  "unique_domains": 654,
}
`;

exports[`/admin/api.php throws not implemented 1`] = `
Object {
  "error": "Method not implemented yet",
  "errorCode": "MethodNotImplemented",
  "query": Object {
    "__testNotImped": "",
  },
}
`;
